econml/dml.py

```diff
@@ -11,14 +11,42 @@
 
 import numpy as np
 import copy
-from .utilities import shape, reshape, ndim, hstack, cross_product, transpose
+from .utilities import (shape, reshape, ndim, hstack, cross_product, transpose,
+                        broadcast_unit_treatments, reshape_treatmentwise_effects, StatsModelsWrapper)
 from sklearn.model_selection import KFold, StratifiedKFold
 from sklearn.linear_model import LinearRegression, LassoCV
 from sklearn.preprocessing import PolynomialFeatures, LabelEncoder, OneHotEncoder
 from sklearn.base import clone, TransformerMixin
 from sklearn.pipeline import Pipeline
 from sklearn.utils import check_random_state
 from .cate_estimator import LinearCateEstimator
+from .inference import StatsModelsInference
+
+
+class _EffectOperation:
+    def __init__(self, pre, d_t, d_y):
+        self._pre = pre
+        self._d_t = d_t
+        self._d_y = d_y
+
+    def apply(self, predict, input, interval, make_const_marginal_effect, *args, **kwargs):
+        if make_const_marginal_effect:
+            # for constant marginal effects, we first need to broadcast each possible treatment
+            # to each input row
+            input = broadcast_unit_treatments(input, self._d_t[0] if self._d_t else 1)
+        preds = predict(self._pre(input), *args, **kwargs)
+        if make_const_marginal_effect:
+            # for constant marginal effects, we need to reshape the results to ensure
+            # that effects are ordered first by outcome, then by treatment
+            if interval:
+                # we want to apply the transformation separately to each half of the interval
+                assert shape(preds)[0] == 2
+                return np.stack([reshape_treatmentwise_effects(pred, self._d_t, self._d_y)
+                                 for pred in preds])
+            else:
+                return reshape_treatmentwise_effects(preds, self._d_t, self._d_y)
+        else:
+            return preds
 
 
 class _RLearner(LinearCateEstimator):
```

Comment:
At the level of the RLearner, we need to enable one more alternative version of inference that is computationally lighter and will be very useful and asymptotically valid: bootstrapping only the second stage, i.e. we fix the nuisance estimates and we only run bootstrap on the final stage estimation. This allows us to avoid re-running the heavy first stage estimation for each bootstrap subsample. 

I suspect this should be easy to implement and enable.

Reply:
#107 was created to address this issue in a future PR.